ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -62,52 +183,97 @@ public BlobId(String id, ClusterMap clusterMap) throws IOException {
   /**
    * Re-constructs existing blobId by deserializing from data input stream
    *
-   * @param stream from which to deserialize the blobid
-   * @param clusterMap of the cluster that the blob id belongs to
+   * @param stream from which to deserialize the blobId.
+   * @param clusterMap of the cluster that the blobId belongs to.
    * @throws IOException
    */
   public BlobId(DataInputStream stream, ClusterMap clusterMap) throws IOException {
     this(stream, clusterMap, false);
   }
 
   /**
-   * Re-constructs existing blobId by deserializing from data input stream. This constructor includes an optional check
-   * that the stream has no more available bytes after reading.
-   *
-   * @param stream from which to deserialize the blobid
-   * @param clusterMap of the cluster that the blob id belongs to
-   * @param ensureFullyRead {@code true} if the stream should have no more available bytes after deserializing the blob
-   *                        ID.
-   * @throws IOException
+   * Size of blobId when it is serialized into bytes.
+   * @return The byte count of the serialized blobId.
    */
-  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead) throws IOException {
-    this.version = stream.readShort();
-    if (version == 1) {
-      partitionId = clusterMap.getPartitionIdFromStream(stream);
-      if (partitionId == null) {
-        throw new IllegalArgumentException("Partition ID cannot be null");
-      }
-      uuid = Utils.readIntString(stream);
-      if (ensureFullyRead && stream.read() != -1) {
-        throw new IllegalArgumentException("Stream should have no more available bytes to read");
-      }
-    } else {
-      throw new IllegalArgumentException("version " + version + " not supported for blob id");
-    }
-  }
-
   public short sizeInBytes() {
-    return (short) (Version_Size_In_Bytes + partitionId.getBytes().length + UUID_Size_In_Bytes + uuid.length());
+    short sizeForBlobIdV1 =
+        (short) (VERSION_FIELD_LENGTH_IN_BYTES + partitionId.getBytes().length + UUID_SIZE_FIELD_LENGTH_IN_BYTES
+            + uuid.length());
+    switch (version) {
+      case BLOB_ID_V1:
+        return sizeForBlobIdV1;
+
+      case BLOB_ID_V2:
+        return (short) (FLAG_FIELD_LENGTH_IN_BYTES + DATACENTER_ID_FIELD_LENGTH_IN_BYTES
+            + ACCOUNT_ID_FIELD_LENGTH_IN_BYTES + CONTAINER_ID_FIELD_LENGTH_IN_BYTES + sizeForBlobIdV1);
+
+      default:
+        throw new IllegalArgumentException("blobId version=" + version + " not supported");
+    }
   }
 
+  /**
+   * Gets the {@link PartitionId} this blob belongs to.
+   * @return The {@link PartitionId}.
+   */
   public PartitionId getPartition() {
     return partitionId;
   }
 
+  /**
+   * Gets the id of the {@link Account} who created this blob. If this information was not available when the
+   * blobId was formed, it will return -1.
```

Comment:
Also one change to be made is changing `LEGACY_XXX_ID` to `UNKNOWN_XXX_ID`. This is because these ids mean no known id information could be obtained when forming the corresponding fields.